### ISSUE TITLE
Syntax highlighting for emacs

### ISSUE THREAD

**Issue by vladaionescu** [earthly/earthly#378]
We already have a [TextMate grammar](https://github.com/earthly/earthly/blob/master/contrib/earthfile-syntax-highlighting/syntaxes/earthfile.tmLanguage.json) for VS Code syntax highlighting. 

We need one for Emacs.

**Comment by AnshMishra2001** [comment]
I am new to Open Source but love to give this a shot. Could you please guide me?

**Comment by vladaionescu** [comment]
Hi @AnshMishra2001 - unfortunately this task does require knowledge of Emacs and syntax highlighting for Emacs.

**Comment by 7vikpeculiar** [comment]
Hi @vladaionescu , I am relatively new to Open Source, but I have experience with emacs, can I give it a shot?

**Comment by vladaionescu** [comment]
Go for it!

**Comment by 7vikpeculiar** [comment]
Hi, I am having some trouble understanding what the TextMate grammar means... Can I please get some help ?

**Comment by TheCoreMan** [comment]
FYI - Opened a project here, might work on it soon

https://github.com/TheCoreMan/earthly-spacemacs-layer

**Comment by vladaionescu** [comment]
Hi @TheCoreMan - I saw that you made some progress over there. Screenshot looks really cool! Do you think that the syntax highlighting is in a good enough state to be used more widely by others in the Earthly community?

**Comment by TheCoreMan** [comment]
@vladaionescu There are a ton of other features I want to include, as you can see in the [README](https://github.com/TheCoreMan/earthly-spacemacs-layer):

> Features (in order of importance)
> - [x] Syntax highlighting
> - [ ] auto-completion (including target names)
> - [ ] Built in pop-up docs
> - [ ] Jump to target (bonus points for other files) with SPC j
> - [ ] call target from within file , e t (, evaluate target)
> - [ ] maybe color successful targets/cached targets? Seems hard

Also, it's not super fun to install yet: Right now one needs to clone -> evaluate buffer instead of just adding a line to the spacemacs config. I'll have to figure out deployment for the layer as well

So... not really? If you want to try and set up some time to co-work on this, I'll definitely be down :)

**Comment by vladaionescu** [comment]
I think even just syntax highlighting could be a big step forward compared to the existing experience. If you figure out the installation process and everything looks good, we could show your project in Earthly install instructions, to help emacs users.

The roadmap is pretty cool. Perhaps `earthly` itself can help with some of these. Some of the functionality could then be used across other IDEs / editors.

For example, if `earthly` had a way to list target names + line numbers for a given file, then that could be used to implement jump to target and auto-completion. For example something like `earthly util list-targets --with-line-numbers ./path/to/Earthfile`. Let me know if you can think of other such standardized utilities that earthly itself could provide.

Regarding pop-up docs: we have this [open proposal for earthlydoc](https://github.com/earthly/earthly/issues/583). It could be interesting if it's integrated with IDEs / editors. Not sure if that's what you had in mind with "pop-up docs".

**Comment by wingyplus** [comment]
HI, 

I build a major mode for emacs [here](https://github.com/wingyplus/earthly-mode/blob/main/earthly-mode.el). I now support syntax highlighting and toggle comment. Here's snapshot test against Earthfile in this repository:

<img width="1440" alt="Screen Shot 2564-08-12 at 11 43 34" src="https://user-images.githubusercontent.com/484530/129140131-f5e198df-b7be-4e00-885c-2b52759ae95f.png">

The syntax highlight references from:

* https://docs.earthly.dev/docs/earthfile.
* https://github.com/earthly/earthly.vim/blob/main/syntax/Earthfile.vim
* https://github.com/TheCoreMan/earthly-spacemacs-layer (Thanks to @TheCoreMan)

And has improvement for variable highlighting, it's support both `$VAR` and `${VAR}`:

<img width="327" alt="Screen Shot 2564-08-12 at 12 00 30" src="https://user-images.githubusercontent.com/484530/129140586-aa778a11-3922-43d3-b883-ef64322d8abd.png">

**Comment by vladaionescu** [comment]
This is awesome!

@wingyplus - would you be able to write some official documentation of how to install it? Here are examples of such instructions for [vim](https://github.com/earthly/earthly.vim) and [sublime](https://github.com/earthly/sublimetext-earthly-syntax).

Once that it's done, we'll need volunteers for testing the installation and the highlighting in their projects.

If all looks good and you're open to it, I'd like to put this in an MPL-licensed repository under the `earthly` GitHub user, giving author credit to both yourself and @TheCoreMan in it.

**Comment by wingyplus** [comment]
> @wingyplus - would you be able to write some official documentation of how to install it? Here are examples of such instructions for vim and sublime.

The README is (update)[https://github.com/wingyplus/earthly-mode]. 

> Once that it's done, we'll need volunteers for testing the installation and the highlighting in their projects.

I think it is almost quite ready. If someone found any bug. PR is welcome!

> If all looks good and you're open to it, I'd like to put this in an MPL-licensed repository under the earthly GitHub user, giving author credit to both yourself and @TheCoreMan in it.

@vladaionescu I really appreciate it. :) 

Do you want me to transfer ownership to the `earthly` org?

**Comment by wingyplus** [comment]
Update from during the weekend:

* The syntax for `FOR .. IN` and `SAVE ARTIFACT .. AS LOCAL` is supported.
* Preparing before submitting to [MELPA](https://github.com/melpa/melpa/blob/master/.github/PULL_REQUEST_TEMPLATE.md)
  * Fix package comment by running `package-lint`. I think we can run it on CI, maybe we can add it on GitHub workflows.
  * Make `checkdoc` happy.

I think it is quite stable enough to let volunteers test a package. :)

**Comment by wingyplus** [comment]
@vladaionescu ping :)

**Comment by vladaionescu** [comment]
Hi @wingyplus - apologies, I was away for a while. This looks really awesome! If you are ok to transfer the project over to the `earthly` org it would be greatly appreciated.

In parallel, if anyone else is able to install and verify that the mode is working as expected and report back to this issue with a comment it would be really useful.

**Comment by duck1123** [comment]
I'm seeing some odd highlighting in places where part of my commands match some of the keywords (expose, use)

I'm not sure I installed it entirely correctly. I copied the earthly-mode directory to my personal directory (I use Prelude) and tried initializing it with

``` emacs-lisp
(add-to-list 'load-path "~/.emacs.d/personal/earthly-mode")
(require 'earthly-mode)
```

but it only provided `earthly-mode/sample-function`. I later evaluated each of the forms in the syntax highlighter.
![Screenshot from 2021-08-23 18-16-36](https://user-images.githubusercontent.com/11774/130526955-75f59fd3-a3d0-479a-bf81-b3c14d4e6fd9.png)

**Comment by wingyplus** [comment]
@vladaionescu Thanks for your reply. I’ll transfer within the day. :)

**Comment by wingyplus** [comment]
@duck1123 Thank you for help testing it. I’ll looking it within the day.

**Comment by duck1123** [comment]
@wingyplus same thing in eye-bleed mode.

Using this file: https://github.com/duck1123/dinsro/blob/develop/Earthfile
![Screenshot from 2021-08-23 20-41-09](https://user-images.githubusercontent.com/11774/130537523-2ac834de-d7f1-4151-87f8-e00b36e1f17c.png)

**Comment by wingyplus** [comment]
@duck1123 thanks i’ll looking it in a few hours

**Comment by wingyplus** [comment]
@duck1123 Now on the master branch, the target name issue is fixed. I'm not sure why the theme that you use highlights the variable partially but it's not happening on the default emacs theme. (The dark theme in the screenshot below is iTerm nord theme) :(

<img width="1030" alt="Screen Shot 2564-08-24 at 12 17 12" src="https://user-images.githubusercontent.com/484530/130559856-e8a8eeba-5886-48f3-aba1-de5cd3fef01e.png">

**Comment by wingyplus** [comment]
@vladaionescu I try to transfer ownership but it requires me to have permission to create a repository on the earthly organization. The error after transfer is:

```
You don’t have the permission to create public repositories on earthly 
```

Or we might try changing to approach such as fork it to earthly organization or push it earthly organization by you. :)

**Comment by wingyplus** [comment]
@vladaionescu another question is do we need to treat `.earth` file extension to enable syntax highlight by default?

**Comment by wingyplus** [comment]
@vladaionescu Thanks for opening the repository. The pull request is now open here https://github.com/earthly/earthly-mode/pull/1. :)

**Comment by wingyplus** [comment]
Since https://github.com/earthly/earthly-mode/pull/1 was merged. The next step is publishing it to MELPA. :)

**Comment by vladaionescu** [comment]
Great! Will you be doing this @wingyplus? You can say that you are a maintainer if they ask.

**Comment by wingyplus** [comment]
Sure. Let me read the submitting instruction and will tell in this issue after I submitting to them. :)

**Comment by wingyplus** [comment]
I'm forgot about changing url to earthly org. Please help review https://github.com/earthly/earthly-mode/pull/2/files. 🙇‍♂️

**Comment by wingyplus** [comment]
I submitted a PR to MELPA here https://github.com/melpa/melpa/pull/7701. Waiting for maintainer to reviewing it.

**Comment by wingyplus** [comment]
Hi folks, earthfile-mode (previous name earthly-mode) is now published to [MELPA](https://melpa.org/#/earthfile-mode). You can now install package directly from emacs. 

If you have a trouble or found a bug, you can report it at https://github.com/earthly/earthly-emacs/issues. 😃 

/cc @vladaionescu

**Comment by vladaionescu** [comment]
Great stuff @wingyplus 👏 🎉 

Would you mind updating the project's readme to add instructions on how to install from MELPA? I'll then use those same instructions on the "[get earthly](https://earthly.dev/get-earthly)" page.

**Comment by wingyplus** [comment]
> Great stuff @wingyplus 👏 🎉
> 
> Would you mind updating the project's readme to add instructions on how to install from MELPA? I'll then use those same instructions on the "[get earthly](https://earthly.dev/get-earthly)" page.

I’ll open a PR within today

**Comment by wingyplus** [comment]
@vladaionescu The readme updated here https://github.com/earthly/earthly-emacs/pull/11.

**Comment by vladaionescu** [comment]
Aaaaand it's live 🎉 

See under **Syntax highlighting**: https://earthly.dev/get-earthly

@duck1123 - would you be able to install using the instructions on the "Get Earthly" page ☝️  and verify that all works as expected?

**Comment by duck1123** [comment]
I don't have use-package to test with, but I can confirm that the package-install instructions work.

**Comment by wingyplus** [comment]
Nice!!

**Comment by vladaionescu** [comment]
Awesome work, everyone! Thank you @wingyplus for putting everything together!

This is now complete. Future work will be tracked as part of the earthly/earthly-emacs repository.
